xskillscore/core/comparative.py

```diff
@@ -0,0 +1,162 @@
+        >>> st = sign_test(f1, f2, o, dim='time')
+        >>> st.sel(results='sign_test').plot()
+        >>> st.sel(results='confidence').plot(c='gray')
+        >>> (-1*st.sel(results='confidence')).plot(c='gray')
+
+        References
+        ----------
+            * DelSole, T., & Tippett, M. K. (2016). Forecast Comparison Based on Random
+              Walks. Monthly Weather Review, 144(2), 615–626. doi: 10/f782pf
+    """
+    # make sure metric is a callable
+    if metric is not None:
+        if isinstance(metric, str):
+            import xskillscore as xs
+
+            if hasattr(xs, metric):
+                metric = getattr(xs, metric)
+            else:
+                raise ValueError(
+                    f'xskillscore.metric could not be derived from {metric}.'
+                )
+        elif not callable(metric):
+            raise ValueError(
+                f'metric needs to be a function/callable or None, found {type(metric)}'
+            )
+
+    if metric:
+        if observation is not None:
+            # Compare the forecasts and observation using metric
+            diff1 = metric(forecast1, observation, dim=[])
+            diff2 = metric(forecast2, observation, dim=[])
+        else:
+            raise ValueError(
+                'observations must be provided when metric is provided', UserWarning
+            )
+    else:  # if metric=None, already evaluated
+        if observation is not None:
+            warnings.warn(
+                'Ignoring provided observation because no metric was provided',
+                UserWarning,
+            )
+        if orientation == 'negative':
+            diff1 = forecast1
+            diff2 = forecast2
+        elif orientation == 'positive':
+            diff1 = -forecast1
+            diff2 = -forecast2
+        if orientation not in ['negative', 'positive']:
+            raise ValueError(
+                '`orientation` requires to be either "positive" or'
+                f'"negative"], found {orientation}.'
+            )
+
+    if categorical:  # ignores orientation and warns if metric provided
+        if metric:
+            warnings.warn('Ignoring provided metric because categorical=True')
+        if observation is not None:
+            diff1 = -1 * (forecast1 == observation)
+            diff2 = -1 * (forecast2 == observation)
+        else:
+            diff1 = ~forecast1
+            diff2 = ~forecast2
+
+    sign_test = (1 * (diff1 < diff2) - 1 * (diff2 < diff1)).cumsum(dim)
+
+    # Estimate 95% confidence interval -----
+    notnan = 1 * (diff1.notnull() & diff2.notnull())
+    N = notnan.cumsum(dim)
+    # z_alpha is the value at which the standardized cumulative Gaussian distributed exceeds alpha
+    confidence = st.norm.ppf(1 - alpha / 2) * xr.ufuncs.sqrt(N)
+    confidence.coords['alpha'] = alpha
+
+    res = xr.concat([sign_test, confidence], dim='results')
+    res['results'] = ['sign_test', 'confidence']
```

Comment:
ideas for a better name than results?

Reply:
we could also return 2 objects, as this is not a metric but rather a test...

Reply:
I don't have a strong opinion here.

There are other functions that we plan to implement in the future that also have multiple outputs that can not be related intuitively along a new dimension. So it would be good to reach a consensus here to minimize future changes. Any thoughts @bradyrx, @raybellwaves?

Reply:
I think many packages like `scipy` and `numpy` will return multiple outputs. E.g. https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.pearsonr.html and https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.linregress.html.

So I prefer two separate outputs. I find that concatenating over custom dimensions/coordinates can sometimes be quite confusing for the user. Probably clearer here to just have the test results and confidence intervals separate.

Reply:
confidence can also be a coordinate